hazelcast/src/main/java/com/hazelcast/cache/impl/CacheMergeRunnable.java

```diff
@@ -0,0 +1,124 @@
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.cache.impl.ICacheService.SERVICE_NAME;
+import static com.hazelcast.config.MergePolicyConfig.DEFAULT_BATCH_SIZE;
+import static com.hazelcast.spi.impl.merge.MergingHolders.createMergeHolder;
+
+class CacheMergeRunnable extends AbstractMergeRunnable<ICacheRecordStore, MergingEntryHolder<Data, Data>> {
+
+    private final CacheService cacheService;
+    private final CacheSplitBrainHandlerService cacheSplitBrainHandlerService;
+
+    CacheMergeRunnable(Map<String, Collection<ICacheRecordStore>> collectedStores,
+                       Map<String, Collection<ICacheRecordStore>> collectedStoresWithLegacyPolicies,
+                       Collection<ICacheRecordStore> backupStores,
+                       CacheSplitBrainHandlerService cacheSplitBrainHandlerService,
+                       NodeEngine nodeEngine) {
+        super(CacheService.SERVICE_NAME, collectedStores, collectedStoresWithLegacyPolicies, backupStores, nodeEngine);
+
+        this.cacheService = nodeEngine.getService(SERVICE_NAME);
+        this.cacheSplitBrainHandlerService = cacheSplitBrainHandlerService;
+    }
+
+    @Override
+    protected void consumeStore(ICacheRecordStore store, BiConsumer<Integer, MergingEntryHolder<Data, Data>> consumer) {
+        int partitionId = store.getPartitionId();
+
+        for (Map.Entry<Data, CacheRecord> entry : store.getReadOnlyRecords().entrySet()) {
+            Data key = entry.getKey();
+            CacheRecord record = entry.getValue();
+            Data dataValue = toData(record.getValue());
+
+            consumer.accept(partitionId, createMergeHolder(key, dataValue, record));
+        }
+    }
+
+    @Override
+    protected void consumeStoreLegacy(ICacheRecordStore recordStore, BiConsumer<Integer, Operation> consumer) {
+        int partitionId = recordStore.getPartitionId();
+        String name = recordStore.getName();
+        CacheMergePolicy mergePolicy = ((CacheMergePolicy) getMergePolicy(name));
+
+        for (Map.Entry<Data, CacheRecord> entry : recordStore.getReadOnlyRecords().entrySet()) {
+            Data key = entry.getKey();
+            CacheRecord record = entry.getValue();
+            CacheEntryView<Data, Data> entryView = new DefaultCacheEntryView(
+                    key,
+                    toData(record.getValue()),
+                    record.getCreationTime(),
+                    record.getExpirationTime(),
+                    record.getLastAccessTime(),
+                    record.getAccessHit());
+
+            consumer.accept(partitionId, new CacheLegacyMergeOperation(name, key, entryView, mergePolicy));
+        }
+    }
+
+    @Override
+    protected InMemoryFormat getInMemoryFormat(String dataStructureName) {
+        return cacheSplitBrainHandlerService.getConfigs().get(dataStructureName).getInMemoryFormat();
+    }
+
+    // TODO why can't we get the configured batchSize from MergePolicyConfig?
```

Comment:
@Donnerbart any ideas?

Reply:
The `CacheConfig` cannot be changed due to `@BinaryInterface` and client compatibility. So we decided to just keep the old merge policy String for `ICache` and use the default value for the batch size. We could add the `MergePolicyConfig` to `CacheSimpleConfig`, but that is not used all the time and `CacheConfig` is always used internally and shared between the members. So there is no real way to add something to the cache configuration. I think this TODO should be removed. The "special configuration" of `ICache` merge policies is already part of my ref manual draft.

Reply:
added a note, removed TODO